File: DataTierGeneratorPlusTester/Cs_BuiltIn/BindingListView.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace Cs_BuiltIn {

    public class BindingListView<T> : BindingList<T>, IBindingListView, IRaiseItemChangedEvents
      {
            private Boolean _isSorted = false;
            private Boolean _isFiltered = false;
            private String _filterString = null;
            private Predicate<T> _filterPredicate = null;
            private ListSortDirection _sortDirection = ListSortDirection.Ascending;
            private PropertyDescriptor _sortProperty = null;
            private ListSortDescriptionCollection _sortDescriptions = new ListSortDescriptionCollection();
            private List<T> _originalCollection = new List<T>();
 
            public BindingListView() : base()
            {
            }
 
            public BindingListView(List<T> list) : base(list)
            {
            }
 
            protected override Boolean SupportsSearchingCore
            {
                  get { return true; }
            }
 
            protected override int FindCore(PropertyDescriptor property, object key)
            {
                  // Simple iteration:
                  for (int i = 0; i < Count; i++)
                  {
                        T item = this[i];
                        if (property.GetValue(item).Equals(key))
                        {
                              return i;
                        }
                  }
                  return -1; // Not found
            }
 
            protected override Boolean SupportsSortingCore
            {
                  get { return true; }
            }
 
            protected override Boolean IsSortedCore
            {
                  get { return _isSorted; }
            }
 
            protected override ListSortDirection SortDirectionCore
            {
                  get { return _sortDirection; }
            }
 
            protected override PropertyDescriptor SortPropertyCore
            {
                  get { return _sortProperty; }
            }
 
            protected override void ApplySortCore(PropertyDescriptor property, ListSortDirection direction)
            {
                  _sortDirection = direction;
                  _sortProperty = property;
                  SortComparer<T> comparer = new SortComparer<T>(property, direction);
                  ApplySortInternal(comparer);
            }
 
            private void ApplySortInternal(SortComparer<T> comparer)
            {
                  if (_originalCollection.Count == 0)
                  {
                        _originalCollection.AddRange(this);
                  }
                  List<T> listRef = this.Items as List<T>;
                  if (listRef == null)
                        return;
                  listRef.Sort(comparer);
                  _isSorted = true;
                  OnListChanged(new ListChangedEventArgs(ListChangedType.Reset, -1));
            }
 
            protected override void RemoveSortCore()
            {
                if (!_isSorted)
                {
                    return;
                }
 
                  Clear();
 
                  foreach (T item in _originalCollection)
                  {
                        Add(item);
                  }
                  _originalCollection.Clear();
                  _sortProperty = null;
                  _sortDescriptions = null;
                  _isSorted = false;
            }
 
            public void ApplySort(ListSortDescriptionCollection sorts)
            {
                  _sortProperty = null;
                  _sortDescriptions = sorts;
                  SortComparer<T> comparer = new SortComparer<T>(sorts);
                  ApplySortInternal(comparer);
            }
 
            public String Filter
            {
                  get { return _filterString; }
                  set
                  {
                        _filterPredicate = null;
                        _filterString = value;
                        _isFiltered = true;
                        UpdateFilter();
                  }
            }
 
            public Predicate<T> FilterPredicate
            {
                  get { return _filterPredicate; }
                  set
                  {
                        _filterString = null;
                        _filterPredicate = value;
                        _isFiltered = true;
                        UpdateFilter();
                  }
            }
 
            public void RemoveFilter()
            {
                if (!_isFiltered)
                {
                    return;
                }

                  _filterString = null;
                  _filterPredicate = null;
                  _isFiltered = false;
 
                  if (_originalCollection.Count < 1000)
                  {
                        Clear();
                        foreach (T item in _originalCollection)
                        {
                              Add(item);
                        }
                        _originalCollection.Clear();
                  }
                  else
                  {
                        this.ClearItems();
                        foreach (T item in _originalCollection)
                        {
                              this.Items.Add(item);
                        }
                        _originalCollection.Clear();
                        if (!_isSorted)
                        {
                            this.OnListChanged(new ListChangedEventArgs(ListChangedType.Reset, 0));
                        }
                  }

                  if (_isSorted)
                  {
                      this.ApplySortCore(_sortProperty, _sortDirection);
                  }
            }
 
            public ListSortDescriptionCollection SortDescriptions
            {
                  get { return _sortDescriptions; }
            }
 
            public Boolean SupportsAdvancedSorting
            {
                  get { return true; }
            }
 
            public Boolean SupportsFiltering
            {
                  get { return true; }
            }
 
            protected virtual void UpdateFilter()
            {
                  if (_originalCollection.Count == 0)
                  {
                        _originalCollection.AddRange(this);
                  }
 
                  if (_filterString != null)
                  {
                        this.UpdateFilterFromString();
                  }
                  else if (_filterPredicate != null)
                  {
                        this.UpdateFilterFromPredicate();
                  }
                  else
                  {
                        this.RemoveFilter();
                  }
            }
 
            private void UpdateFilterFromString()
            {
                  int equalsPos = _filterString.IndexOf('=');
                  // Get property name
                  String propName = _filterString.Substring(0, equalsPos).Trim();
                  // Get filter criteria
                  String criteria = _filterString.Substring(equalsPos + 1, _filterString.Length - equalsPos - 1).Trim();
                  // Strip leading and trailing quotes
                  if (criteria.Contains("\"") || criteria.Contains("'"))
                  {
                        criteria = criteria.Substring(1, criteria.Length - 2);
                  }
                  // Get a property descriptor for the filter property
                  PropertyDescriptor propDesc = TypeDescriptor.GetProperties(typeof(T))[propName];
                  List<T> currentCollection = new List<T>(this);
                  Clear();
                  foreach (T item in currentCollection)
                  {
                        object value = propDesc.GetValue(item);
                        if (value.ToString() == criteria)
                        {
                              Add(item);
                        }
                  }
            }
 
            private void UpdateFilterFromPredicate()
            {
                  this.Items.Clear();
                  foreach (T item in _originalCollection)
                  {
                        if (this.FilterPredicate(item))
                        {
                              this.Items.Add(item);
                        }
                  }
 
                  if (_isSorted)
                        this.ApplySortCore(_sortProperty, _sortDirection);
                  else
                        this.OnListChanged(new ListChangedEventArgs(ListChangedType.Reset, 0));
            }
 
            public new Boolean AllowNew //use new because property is expected to be AllowNew and base is not virtual 
            {
                  get { return CheckReadOnly(); }
            }
 
            public new Boolean AllowRemove //use new because property is expected to be AllowRemove and base is not virtual 
            {
                  get { return CheckReadOnly(); }
            }

            private Boolean CheckReadOnly()
            {
                  if (_isSorted || _isFiltered)
                  {
                        return false;
                  }
                  else
                  {
                        return true;
                  }
            }
 
            protected override void InsertItem(int index, T item)
            {
                  foreach (PropertyDescriptor propDesc in TypeDescriptor.GetProperties(item))
                  {
                        if (propDesc.SupportsChangeEvents)
                        {
                              propDesc.AddValueChanged(item, OnItemChanged);
                        }
                  }
                  base.InsertItem(index, item);
            }
 
            protected override void RemoveItem(int index)
            {
                  T item = Items[index];
                 PropertyDescriptorCollection propDescs = TypeDescriptor.GetProperties(item);

                  foreach (PropertyDescriptor propDesc in propDescs)
                  {
                        if (propDesc.SupportsChangeEvents)
                        {
                              propDesc.RemoveValueChanged(item, OnItemChanged);
                        }
                  }
                  base.RemoveItem(index);
            }
 
            protected void OnItemChanged(Object sender, EventArgs args)
            {
                  int index = Items.IndexOf((T)sender);
                  OnListChanged(new ListChangedEventArgs(ListChangedType.ItemChanged, index));
            }
 
            public Boolean RaisesItemChangedEvents
            {
                  get { return true; }
            }
 
      }

}

File: DataTierGeneratorPlusTester/VB_BuiltIn/BindingListView.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel

'NameSpace VB_BuiltIn

Public Class BindingListView(Of T)
    Inherits BindingList(Of T)
    Implements IRaiseItemChangedEvents, IBindingListView


    Private _isSorted As Boolean = False
    Private _isFiltered As Boolean = False
    Private _filterString As String = Nothing
    Private _filterPredicate As Predicate(Of T) = Nothing
    Private _sortDirection As ListSortDirection = ListSortDirection.Ascending
    Private _sortProperty As PropertyDescriptor = Nothing
    Private _sortDescriptions As ListSortDescriptionCollection = New ListSortDescriptionCollection()
    Private _originalCollection As List(Of T) = New List(Of T)()

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(ByVal list As List(Of T))
        MyBase.New(list)
    End Sub

    Protected Overrides ReadOnly Property SupportsSearchingCore() As Boolean
        Get
            Return True
        End Get
    End Property

    Protected Overrides Function FindCore(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer

        ' Simple iteration:

        For i As Integer = 0 To Count - 1
            Dim item As T = Me.Items(i)
            If ([property].GetValue(item).Equals(key)) Then
                Return i
            End If
        Next
        Return -1 ' Not found

    End Function

    Protected Overrides ReadOnly Property SupportsSortingCore() As Boolean
        Get
            Return True
        End Get
    End Property

    Protected Overrides ReadOnly Property IsSortedCore() As Boolean
        Get
            Return _isSorted
        End Get
    End Property

    Protected Overrides ReadOnly Property SortDirectionCore() As ListSortDirection
        Get
            Return _sortDirection
        End Get
    End Property

    Protected Overrides ReadOnly Property SortPropertyCore() As PropertyDescriptor
        Get
            Return _sortProperty
        End Get
    End Property

    Protected Overrides Sub ApplySortCore(ByVal [property] As PropertyDescriptor, ByVal direction As ListSortDirection)
        _sortDirection = direction
        _sortProperty = [property]
        Dim Comparer As SortComparer(Of T) = New SortComparer(Of T)([property], direction)
        ApplySortInternal(Comparer)
    End Sub

    Private Sub ApplySortInternal(ByVal comparer As SortComparer(Of T))

        If (_originalCollection.Count = 0) Then
            _originalCollection.AddRange(Me)
        End If
        Dim listRef As List(Of T) = CType(Me.Items, List(Of T))
        If (listRef Is Nothing) Then
            Return
        End If
        listRef.Sort(comparer)
        _isSorted = True
        OnListChanged(New ListChangedEventArgs(ListChangedType.Reset, -1))
    End Sub

    Protected Overrides Sub RemoveSortCore()

        If (Not _isSorted) Then
            Return
        End If

        Clear()
        For Each item As T In _originalCollection
            Add(item)
        Next
        _originalCollection.Clear()
        _sortProperty = Nothing
        _sortDescriptions = Nothing
        _isSorted = False
    End Sub

    Public Sub ApplySort(ByVal sorts As ListSortDescriptionCollection) Implements IBindingListView.ApplySort
        _sortProperty = Nothing
        _sortDescriptions = sorts
        Dim Comparer As SortComparer(Of T) = New SortComparer(Of T)(sorts)
        ApplySortInternal(Comparer)
    End Sub

    Public Property Filter() As String Implements IBindingListView.Filter

        Get
            Return _filterString
        End Get

        Set(ByVal value As String)
            _filterPredicate = Nothing
            _filterString = value
            _isFiltered = True
            UpdateFilter()
        End Set

    End Property

    Public Property FilterPredicate() As Predicate(Of T)

        Get
            Return _filterPredicate
        End Get

        Set(ByVal value As Predicate(Of T))
            _filterString = Nothing
            _filterPredicate = value
            _isFiltered = True
            UpdateFilter()
        End Set

    End Property

    Public Sub RemoveFilter() Implements IBindingListView.RemoveFilter

        If (Not _isFiltered) Then
            Return
        End If
        _filterString = Nothing
        _filterPredicate = Nothing
        _isFiltered = False

        If (_originalCollection.Count < 1000) Then

            Clear()
            For Each item As T In _originalCollection
                Add(item)
            Next
            _originalCollection.Clear()

        Else

            Me.ClearItems()
            For Each item As T In _originalCollection

                Me.Items.Add(item)
            Next
            _originalCollection.Clear()
            If (Not _isSorted) Then
                Me.OnListChanged(New ListChangedEventArgs(ListChangedType.Reset, 0))
            End If
        End If

        If (_isSorted) Then
            Me.ApplySortCore(_sortProperty, _sortDirection)
        End If
    End Sub

    Public ReadOnly Property SortDescriptions() As ListSortDescriptionCollection Implements IBindingListView.SortDescriptions
        Get
            Return _sortDescriptions
        End Get
    End Property

    Public ReadOnly Property SupportsAdvancedSorting() As Boolean Implements IBindingListView.SupportsAdvancedSorting
        Get
            Return True
        End Get
    End Property

    Public ReadOnly Property SupportsFiltering() As Boolean Implements IBindingListView.SupportsFiltering
        Get
            Return True
        End Get
    End Property

    Protected Overridable Sub UpdateFilter()

        If (_originalCollection.Count = 0) Then
            _originalCollection.AddRange(Me)
        End If

        If (Not _filterString Is Nothing) Then
            Me.UpdateFilterFromString()
        ElseIf (Not _filterPredicate Is Nothing) Then
            Me.UpdateFilterFromPredicate()
        Else
            Me.RemoveFilter()
        End If

    End Sub

    Private Sub UpdateFilterFromString()

        Dim equalsPos As Integer = _filterString.IndexOf("=")
        ' Get property name
        Dim propName As String = _filterString.Substring(0, equalsPos).Trim()
        ' Get filter criteria
        Dim criteria As String = _filterString.Substring(equalsPos + 1, _filterString.Length - equalsPos - 1).Trim()
        ' Strip leading and trailing quotes
        If (criteria.Contains("""") Or criteria.Contains("'")) Then
            criteria = criteria.Substring(1, criteria.Length - 2)
        End If
        ' Get a property descriptor for the filter property
        Dim x As T 'GetProperties won't take type T; give it a variable of type T
        Dim propDesc As PropertyDescriptor = TypeDescriptor.GetProperties(x)(propName)
        Dim currentCollection As List(Of T) = New List(Of T)(Me)
        Clear()
        For Each item As T In currentCollection

            Dim value As Object = propDesc.GetValue(item)
            If (value.ToString() = criteria) Then
                Add(item)
            End If
        Next
    End Sub

    Private Sub UpdateFilterFromPredicate()

        Me.Items.Clear()
        For Each item As T In _originalCollection
            If (Me.FilterPredicate(item)) Then
                Me.Items.Add(item)
            End If
        Next

        If (_isSorted) Then
            Me.ApplySortCore(_sortProperty, _sortDirection)
        Else
            Me.OnListChanged(New ListChangedEventArgs(ListChangedType.Reset, 0))
        End If
    End Sub

    Public Shadows ReadOnly Property AllowNew() As Boolean 'use Shadows because property is expected to be AllowNew and base is not virtual 
        Get
            Return CheckReadOnly()
        End Get
    End Property


    Public Shadows ReadOnly Property AllowRemove() As Boolean 'use Shadows because property is expected to be AllowRemove and base is not virtual 
        Get
            Return CheckReadOnly()
        End Get
    End Property

    Private Function CheckReadOnly() As Boolean

        If (_isSorted Or _isFiltered) Then
            Return False
        Else
            Return True
        End If

    End Function

    Protected Overrides Sub InsertItem(ByVal index As Integer, ByVal item As T)

        For Each propDesc As PropertyDescriptor In TypeDescriptor.GetProperties(item)
            If (propDesc.SupportsChangeEvents) Then
                propDesc.AddValueChanged(item, AddressOf OnItemChanged)
            End If
        Next
        MyBase.InsertItem(index, item)

    End Sub

    Protected Overrides Sub RemoveItem(ByVal index As Integer)

        Dim item As T = Items(index)
        Dim propDescs As PropertyDescriptorCollection = TypeDescriptor.GetProperties(item)

        For Each propDesc As PropertyDescriptor In propDescs
            If (propDesc.SupportsChangeEvents) Then
                propDesc.RemoveValueChanged(item, AddressOf OnItemChanged)
            End If
        Next
        MyBase.RemoveItem(index)

    End Sub

    Protected Sub OnItemChanged(ByVal sender As Object, ByVal args As EventArgs)
        Dim index As Integer = Items.IndexOf(CType(sender, T))
        OnListChanged(New ListChangedEventArgs(ListChangedType.ItemChanged, index))
    End Sub

    Public ReadOnly Property RaisesItemChangedEvents() As Boolean
        Get
            Return True
        End Get
    End Property

End Class

'End NameSpace
